docs/downloader.rst

```diff
@@ -0,0 +1,133 @@
+  - ``snapshot_version``: int(1) currently
+  - ``name``: the name of this snapshot (a mangled relative path)
+  - ``author``: a dict containing:
+    - ``name``: arbitrary name
+    - ``verify_key``: base64-encoded public key of the author
+  - ``parents``: a list of immutable Capability-strings of any parent Snapshots
+  - (not yet) additionally, in the Tahoe metadata for this metadata-capability is a ``magic_folder`` dict with the following keys:
+    - ``author_signature``: base64-encoded signature which signs the content-Capability, metadata-Capability and name
+
+If there are zero parents this is the first version of a file.
+If there is one parent it is a modification.
+If there are two or more parents this version is the resolution of a conflict.
+
+When reading the code, there is a method ``create_snapshot_from_capability`` which downloads a capability-string from Tahoe and returns a ``RemoteSnapshot`` instance.
+
+
+General Operation
+-----------------
+
+The ``DownloaderService`` is responsible for deciding which Snapshots to download.
+The capability strings of Snapshots to download are given to ``RemoteSnapshotCacheService`` which downloads a Snapshot and all its relevant parents into memory.
+Once a Snapshot is downloaded it is passed to the ``MagicFolderUpdaterService`` which stages the actual content and determins if the change is an "overwrite" or a "conflict" and changes the local filesystem accordingly.
+
+Downloading new Snapshots from other participants ultimately causes changes to the local filesystem (in the magic-folder).
+
+(Not implemented). It may be the case that one or more parent Snapshots are missing.
+In this case, we cannot cache or otherwise see those Snapshots and must take local action accordingly; see the section on conflicts for more.
+
+
+What Snapshots to Download
+--------------------------
+
+Each configured magic-folder has a ``collective_dircap`` which is a Tahoe-LAFS Directory Capability ("dircap") for the list of participants.
+If this dircap is writable then this device is the administrator (and the only device which can modify the participants).
+
+In either case, the capability can be downloaded.
+It will be a Tahoe-LAFS directory containing a series of sub-directories; these are the participants.
+The directory name is the Participant name and points at a directory read-capability where all the files in their magic-folder are stored.
+This directory is known as the "Personal DMD" (where DMD stands for "Distributed Mutable Directory").
+
+The entries in a user's Personal DMD are flat (no subdirectories) and point a (mangled) relative path-name to a Snapshot.
+
+So, Snapshots to download are discovered by:
+
+- reading the Collective DMD
+- for each user in it:
+  - read their Personal DMD
+  - for each entry in the Personal DMD:
+    - queue the Snapshot for caching, including all parents (see note)
+    - after download, queue the Snapshot for updates
```

Comment:
I'm curious why we handle each file of each participant-device individually, rather than for each file that is from any participant-device, handle all the remote snapshots corresponding to that file?

Reply:
Is there a reason to prefer that order?

Reply:
That way, we only deal with any given file once, rather potentially multiple times, if multiple other participants have updated the file. The current order could potentially cause us to download several versions of a multi-gb file, only to overwrite it later in the process. It also means we may detect conflict in a given file later than we otherwise (after having already download one version an moved it in place, rather than before downloading any file.

Reply:
Yes, I can imagine cases when there is a more-efficient order to do operations in.

I think it makes sense as a "later" / follow-up "optimization" ticket; there are other optimizations we could make and it's way harder to hit any "inefficient" scenario with just two participants anyway.

However, note that "caching the snapshot" does _not_ include downloading the content pointer; that only happens in the "do updates" phase.

Reply:
> Yes, I can imagine cases when there is a more-efficient order to do operations in.

I don't think it is *just* an optimization, but It also has implications for correctness.

> I think it makes sense as a "later" / follow-up "optimization" ticket;

I disagree. I think this has enough impact on behavior, and reasoning about whether our handling (or not) of conflicts is correct that we should deal with now. This is a blocking concern.

> it's way harder to hit any "inefficient" scenario with just two participants anyway.

We'll have (or should have) multiple participants as soon as somebody recovers twice. If we don't, they may lose data if recovery was not complete on the intermediate device.

> However, note that "caching the snapshot" does not include downloading the content pointer; that only happens in the "do updates" phase.

The relevant part isn't "cache the snapshot", but "queue the snapshot for update". We will queue multiple snapshots for the same file, if it is present in multiple remotes. It doesn't really matter which order we loop through participants, but the order we queue the snapshots we find (and how much they are chunked).

Reply:
Okay, what correctness issues might there be from processing order?